### PR TITLE
Refactor implementation of RustSemaphore

### PR DESCRIPTION
**Pull request by RusherRG**
## Description

Fixes https://github.com/Lind-Project/lind_project/issues/321

 

Refactored the implementation of RustSemaphore struct because previously it was using `Atomicu32` to store the value variable but it was causing a [Time-of-check to time-of-use](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) bug. If there were 2 processes waiting for the lock and a 3rd process released the lock then in such scenarios both the processes that were waiting used to get the lock which shouldn't be happening. 

This bug was fixed by using `Mutex<u32>` instead of `Atomicu32` making sure that only one thread/process can read/update the value of the variable at any given time.

### Type of change

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

- `lind_project/tests/test_cases/coreutils_suite/shm_sem_t.sh`
- `lind_project/tests/test_cases/coreutils_suite/shm_sem_fork_t.sh`

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, lind-project)

### CODE REVIEW

File: src/interface/misc.rs [hunk 1]
```diff
@@ -381,70 +381,58 @@ pub struct RustSemaphore {
...
-            Err(_) => false,
+    pub fn lock(&self) {
+        loop {
+            let mut value = self.value.lock();
+            if *value == 0 {
+                interface::lind_yield();
+                continue;
+            }
+                
+            *value = if *value > 0 { *value - 1 } else { 0 };
```
Comment (rennergade) [inline src/interface/misc.rs:397]: this should be an if/else clause, if you do it that way you dont need the continue

File: src/interface/misc.rs [hunk 2]
```diff
@@ -381,70 +381,58 @@ pub struct RustSemaphore {
...
-
-        match result {
-            Ok(_) => true,
-            Err(_) => false,
+        let mut value = self.value.lock();
+        if *value == 0 {
+            return false;
         }
+
+        *value = if *value > 0 { *value - 1 } else { 0 };
```
Comment (rennergade) [inline src/interface/misc.rs:417]: again should be if/else

**Review by rennergade — CHANGES_REQUESTED** [review]
A few minor changes requested. If you can add a few comments explaining the reasoning that would be good too.

**Comment by RusherRG** [comment]
> A few minor changes requested. If you can add a few comments explaining the reasoning that would be good too.

Yeah, I'll make those changes. Also, do you mean adding comments to the code?

File: src/interface/misc.rs [hunk 3]
```diff
@@ -381,70 +381,58 @@ pub struct RustSemaphore {
...
-        });
-
-        match result {
-            Ok(_) => true,
-            Err(_) => false,
-        }
-
+    pub fn unlock(&self) {
+        let mut value = self.value.lock();
+        *value = if *value < (SEM_VALUE_MAX - 1) { *value + 1 } else { SEM_VALUE_MAX };
```
Comment (Yaxuan-w) [inline src/interface/misc.rs:404]: According to the [sem_post](https://man7.org/linux/man-pages/man3/sem_post.3.html) man page, we may need to report error=EOVERFLOW when the value larger than SEM_VALUE_MAX?

Comment (rennergade) [reply src/interface/misc.rs:404]: Good catch we probably need to return an error here

Comment (RusherRG) [reply src/interface/misc.rs]: Thanks for pointing this out. I've made the relevant changes.

**Review by rennergade — APPROVED** [review]
Looks good now. Approved!

**Pull request merged**